File: app/src/main/java/com/leo/androidutil/ui/LocationActivity.java
package com.leo.androidutil.ui;

import android.Manifest;
import android.location.Address;
import android.location.Location;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.text.InputFilter;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.leo.androidutil.R;
import com.leo.commonutil.calendar.DateUtil;
import com.leo.commonutil.asyn.WeakHandler;
import com.leo.commonutil.enume.UnitTime;
import com.leo.commonutil.location.OnLocationCallback;
import com.leo.commonutil.location.SystemLocationUtil;
import com.leo.system.LogUtil;
import com.tbruyelle.rxpermissions2.RxPermissions;

/**
 * 系统定位demo
 */
public class LocationActivity extends BaseActivity implements OnLocationCallback, View.OnClickListener {
    private static final String TAG = LocationActivity.class.getSimpleName();
    private TextView mResultTv;
    private Button mPermissionBtn;
    private Button mStopBtn;
    private WeakHandler weakHandler = new WeakHandler();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_location);
        initActionBar();
        mResultTv = findViewById(R.id.resultTv);
        mPermissionBtn = findViewById(R.id.permissionBtn);
        mPermissionBtn.setOnClickListener(this);
        mStopBtn = findViewById(R.id.stopBtn);
        mStopBtn.setOnClickListener(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (null != weakHandler) {
            weakHandler.removeCallbacksAndMessages(null);
        }
        SystemLocationUtil.getInstance().stop();
    }

    @Override
    protected void initActionBar() {
        super.initActionBar();
        ActionBar actionBar = getSupportActionBar();
        if (null == actionBar) {
            return;
        }
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setTitle("系统定位");
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void location() {
        RxPermissions rxPermissions = new RxPermissions(this);
        rxPermissions.request(Manifest.permission.ACCESS_FINE_LOCATION)
                .subscribe(aBoolean -> {
                    if (aBoolean) {
                        String bestProvider = SystemLocationUtil.getInstance().getBestProvider();
                        SystemLocationUtil.getInstance().start(bestProvider,
                                10000, this);
                    } else {
                        LogUtil.e(TAG, "lack of permission");
                    }
                }, throwable -> {
                    LogUtil.e(TAG, throwable.toString());
                });
    }

    @Override
    public void onLocationChanged() {

    }

    @Override
    public void onLocationReverGeoResult() {
        weakHandler.post(() -> {
            SystemLocationUtil locationUtil = SystemLocationUtil.getInstance();
            Address addressBean = locationUtil.getAddressBean();
            if (null == addressBean) {
                return;
            }
            mResultTv.setText(DateUtil.format(UnitTime.MILLIONSECOND, System.currentTimeMillis(),
                    DateUtil.DATA_YMDHMS));
            mResultTv.append("\n");
            mResultTv.append(locationUtil.getAddressStr(addressBean));
            mResultTv.append("\n");
            Location wgs84 = locationUtil.getLocationWGS84();
            if (null != wgs84) {
                mResultTv.append("国标系 (wgs84)：");
                mResultTv.append("\n");
                mResultTv.append(wgs84.getLatitude() + " - " + wgs84.getLongitude());
                mResultTv.append("\n");
            }
            Location gcj02 = locationUtil.getLocationGCJ02();
            if (null != gcj02) {
                mResultTv.append("火星坐标系 (GCJ-02)：");
                mResultTv.append("\n");
                mResultTv.append(gcj02.getLatitude() + " - " + gcj02.getLongitude());
                mResultTv.append("\n");
            }
            Location bd09LL = locationUtil.getLocationBD09LL();
            if (null != bd09LL) {
                mResultTv.append("百度坐标系 (BD-09)：");
                mResultTv.append("\n");
                mResultTv.append(bd09LL.getLatitude() + " - " + bd09LL.getLongitude());
                mResultTv.append("\n");
            }
            mResultTv.append("反GEO：");
            mResultTv.append("\n");
            mResultTv.append(addressBean.getLatitude() + " - " + addressBean.getLongitude());
        });
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.permissionBtn:
                location();
                break;
            case R.id.stopBtn:
                SystemLocationUtil.getInstance().stop();
                break;
        }
    }
}

File: app/src/main/java/com/leo/androidutil/ui/LocationActivity.kt
package com.leo.androidutil.ui

import android.Manifest
import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.leo.androidutil.R
import com.leo.androidutil.databinding.ActivityLocationBinding
import com.leo.androidutil.viewmodels.LocationModel
import com.leo.commonutil.calendar.DatePresetFormat.DATA_YMDHMS
import com.leo.commonutil.calendar.DateUtil
import com.leo.commonutil.enume.UnitTime
import com.leo.commonutil.location.OnLocationCallback
import com.leo.commonutil.location.SystemLocationUtil
import com.leo.lib_permission.annotations.PermissionApply
import com.leo.lib_permission.annotations.PermissionRefused
import com.leo.lib_permission.annotations.PermissionRefusedForever
import com.leo.system.log.XLog

/**
 * 系统定位demo
 */
class LocationActivity : BaseActivity(), OnLocationCallback, View.OnClickListener {
    private lateinit var mBinding: ActivityLocationBinding
    private lateinit var mModel: LocationModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_location)
//        mModel = ViewModelProviders.of(this).get(LocationModel::class.java)
        mModel = ViewModelProvider(this).get(LocationModel::class.java)
        initView()
        initData()
    }

    @SuppressLint("MissingPermission")
    override fun onPause() {
        super.onPause()
        SystemLocationUtil.getInstance().stop()
    }

    override fun initActionBar() {
        super.initActionBar()
        val actionBar = supportActionBar ?: return
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.title = "系统定位"
    }

    private fun initView() {
        mBinding.permissionBtn.setOnClickListener(this)
        mBinding.stopBtn.setOnClickListener(this)
    }

    private fun initData() {
        mModel.mUpdateTime.observe(this, Observer {
            XLog.d(TAG, "数据变化回调:${it}")
            val locationUtil = SystemLocationUtil.getInstance()
            val addressBean = locationUtil.addressBean ?: return@Observer
            mBinding.resultTv.run {
                text = it
                append("\n")
                append(locationUtil.getAddressStr(addressBean))
                append("\n")
                locationUtil.locationWGS84?.let {
                    append("国标系 (wgs84)：\n${it.latitude} - ${it.longitude}\n")
                }
                locationUtil.locationGCJ02?.let {
                    append("火星坐标系 (GCJ-02)：\n${it.latitude} - ${it.longitude}\n")
                }
                locationUtil.locationBD09LL?.let {
                    append("百度坐标系 (BD-09)：\n${it.latitude} - ${it.longitude}\n")
                }
                append("反GEO：\n${addressBean.latitude} - ${addressBean.longitude}")
            }
        })
    }

    @SuppressLint("MissingPermission")
    @PermissionApply(permissions = [Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION],
            requestCode = 1001)
    fun location() {
        XLog.e(TAG, "授权成功")
        val bestProvider = SystemLocationUtil.getInstance().bestProvider
        SystemLocationUtil.getInstance().start(bestProvider,
                10000, this)
    }

    @PermissionRefused(requestCode = 1001)
    fun onPermissionRefused(s: Array<String?>) {
        XLog.e(TAG, "lack of permission")
    }

    @PermissionRefusedForever(requestCode = 1001)
    fun onPermissionRefusedForever(s: Array<String?>) {
        XLog.e(TAG, "lack of permission")
    }

    override fun onLocationChanged() {
    }

    override fun onLocationReverGeoResult() {
        val locationUtil = SystemLocationUtil.getInstance()
        locationUtil.addressBean ?: return
        val timeFormat = DateUtil.format(UnitTime.MILLIONSECOND, System.currentTimeMillis(), DATA_YMDHMS)
        mModel.mUpdateTime.postValue(timeFormat)
    }

    @SuppressLint("MissingPermission")
    override fun onClick(v: View) {
        when (v.id) {
            R.id.permissionBtn -> location()
            R.id.stopBtn -> SystemLocationUtil.getInstance().stop()
        }
    }
}
